### PR TITLE
SmolVLM

### PR DESCRIPTION
**Pull request by andimarafioti**
## Preparing the Article

You're not quite done yet, though. Please make sure to follow this process (as documented [here](https://github.com/huggingface/blog/tree/main?tab=readme-ov-file#how-to-write-an-article-)):

- [x] Add an entry to [_blog.yml](https://github.com/huggingface/blog/blob/main/_blog.yml).
- [x] Add a thumbnail. There are no requirements here, but there is a [template](https://github.com/huggingface/blog?tab=readme-ov-file#how-to-get-a-nice-responsive-thumbnail) if it's helpful.
- [x] Check you use a short title and blog path.
- [x] Upload any additional assets (such as images) to the Documentation Images [repo](https://huggingface.co/datasets/huggingface/documentation-images/tree/main/blog). This is to reduce bloat in the GitHub base repo when cloning and pulling. Try to have small images to avoid a slow or expensive user experience.
- [x] Add metadata (such as authors) to your `md` file. You can also specify `guest` or `org` for the authors.
- [x] Ensure the publication date is correct.
- [x] Preview the content. A quick way is to paste the markdown content in https://huggingface.co/new-blog. **Do not click publish**, this is just a way to do an early check.

Here is an example of a complete PR: https://github.com/huggingface/blog/pull/2382

## Getting a Review

Please make sure to get a review from someone on your team or a co-author.
Once this is done and once all the steps above are completed, you should be able to merge.
There is no need for additional reviews if you and your co-authors are happy and meet all of the above.

Feel free to add @pcuenca as a reviewer if you want a final check. Keep in mind he'll be biased toward light reviews
(e.g., check for proper metadata) rather than content reviews unless explicitly asked.

### CODE REVIEW

File: smolvlm.md [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+
+## Fine-tuning
+
+You can fine-tune SmolVLM using transformers and apply alignment techniques using TRL 🚀
+
+We provide a notebook to fine-tune it on the VQAv2 dataset, optionally using  LoRA, QLoRA or full fine-tuning. In the notebook, you can find some tricks to save up even more memory and have a larger batch size to fit SmolVLM inside consumer GPUs, like L4, for training. With batch sizes of 4, 8-bit loading with QLoRA and gradient checkpointing we can fine-tune in L4, and it consumes around ~16 GBs of VRAM. This makes it possible to fine-tune your SmolVLM using Colab! You can play around with the parameters to get a nice point in training duration-memory trade-off. 
+
+SmolVLM also comes with TRL integration so you can apply Direct Preference Optimization (DPO) easily through the CLI. Get started by running pip install trl accelerate and then run the following command to fine-tune on [RLAIF-V] (https://huggingface.co/datasets/HuggingFaceH4/rlaif-v_formatted) dataset.
```
Comment (kashif) [inline smolvlm.md:239]:
```suggestion
SmolVLM also comes with TRL integration so you can apply Direct Preference Optimization (DPO) easily through the CLI. Get started by running `pip install trl accelerate peft` and then run the following command to fine-tune on [RLAIF-V] (https://huggingface.co/datasets/HuggingFaceH4/rlaif-v_formatted) dataset: 
```

File: smolvlm.md [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+
+``` bash
+accelerate launch --config_file examples/accelerate_configs/multi_gpu.yaml examples/scripts/dpo_vlm.py  \\
+  --dataset_name HuggingFaceH4/rlaif-v_formatted --model_name_or_path HuggingFaceTB/SmolVLM-Instruct \\
+  --per_device_train_batch_size 8 --gradient_accumulation_steps 32 --dataset_num_proc 32 \\
+  --output_dir dpo_smolvlm_rlaif-v --bf16 --torch_dtype bfloat16 --use_peft --lora_target_modules=all-linear 
+```
+
```
Comment (kashif) [inline smolvlm.md:247]:
```suggestion
A detailed tutorial on preference tuning vision-based LLM can be found here: [dpo_vlm](https://huggingface.co/blog/dpo_vlm).
```

File: smolvlm.md [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+SmolVLM also comes with TRL integration so you can apply Direct Preference Optimization (DPO) easily through the CLI. Get started by running `pip install trl accelerate peft` and then run the following command to fine-tune on [RLAIF-V] (https://huggingface.co/datasets/HuggingFaceH4/rlaif-v_formatted) dataset: 
+
+``` bash
+accelerate launch --config_file examples/accelerate_configs/multi_gpu.yaml examples/scripts/dpo_vlm.py  \\
+  --dataset_name HuggingFaceH4/rlaif-v_formatted --model_name_or_path HuggingFaceTB/SmolVLM-Instruct \\
+  --per_device_train_batch_size 8 --gradient_accumulation_steps 32 --dataset_num_proc 32 \\
+  --output_dir dpo_smolvlm_rlaif-v --bf16 --torch_dtype bfloat16 --use_peft --lora_target_modules=all-linear 
+```
+A detailed tutorial on preference tuning vision-based LLM can be found here: [dpo_vlm](https://huggingface.co/blog/dpo_vlm).
```
Comment (kashif) [inline smolvlm.md:247]:
```suggestion
The resulting LoRA adapter weights are [SmolVLM-Instruct-DPO](https://huggingface.co/HuggingFaceTB/SmolVLM-Instruct-DPO). A detailed tutorial on preference tuning vision-based LLM can be found here: [dpo_vlm](https://huggingface.co/blog/dpo_vlm).
```

File: smolvlm.md [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+
+For SmolVLM, we closely followed the architecture from Idefics3, to the point that we use the same implementation in transformers. There are, however a few key differences:
+ - We replaced Llama 3.1 8B with SmolLM2 1.7B as the language backbone.
+ - We more aggressively compress the patched information by reducing the information 9x using the pixel shuffle strategy, compared to 4x with idefics3.
+ - We use patches of 384*384, instead of 364x364, because 384 is divisible by 3, which is necessary for our pixel shuffle strategy to work.
+ - For this, we change the vision backbone to use shape-optimized SigLIP with patches of 384x384 pixels and inner patches of 14x14.
+
+
+```
+All the models in this release are built on the shape-optimized SigLIP as image encoder and SmolLM2 for text decoder part, trained on datasets with Apache 2.0 license. We release the SmolVLM checkpoints under the Apache 2.0 license.
```
Comment (merveenoyan) [inline smolvlm.md:114]: why is thi sin code brackets?

**Review by mfarre — APPROVED** [review]

File: smolvlm.md [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+- user: andito
+- user: merve
+- user: mfarre
+- user: eliebak
+- user: pcuenq
+---
+
+## TLDR
+
+This blog post introduces SmolVLM a 2B SOTA VLM given its memory footprint. SmolVLM is small, fast, memory-efficient, and fully open-source. All model checkpoints, VLM datasets, training recipes and tools are released under the Apache 2.0 license.
```
Comment (pcuenca) [inline smolvlm.md:14]:
```suggestion
This blog post introduces SmolVLM, a 2B VLM, SOTA for its memory footprint. SmolVLM is small, fast, memory-efficient, and fully open-source. All model checkpoints, VLM datasets, training recipes and tools are released under the Apache 2.0 license.
```

File: smolvlm.md [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+<img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/smolvlm_ecosystem.png" width="800" height="auto" alt="Image description">
+
+
+## What is SmolVLM?
+
+This year has seen a boom in multimodal AI with many large vision language models released. The trends were to initially scale up compute, later scale up the data diversity by generating synthetic data with large models, and, recently, scale down to make these models more efficient. Small open models allow local deployment to browser or edge devices, cut inference costs, and enable user customization. Some notable examples of these models include PaliGemma 3B, moondream2, and Qwen2VL.
+
+In this blog post, we introduce SmolVLM, a new family of 2B small vision language models that can be used commercially and deployed to smaller local setups, with completely open training pipelines. 
+
+We release three models: SmolVLM-Base, which can be used for downstream fine-tuning, SmolVLM-Synthetic, the fine-tuned variant on synthetic data, and SmolVLM Instruct, the fine-tuned instruction variant, which can be used out of the box for interactive end-user applications.
```
Comment (pcuenca) [inline smolvlm.md:25]:
```suggestion
We release three models: SmolVLM-Base, which can be used for downstream fine-tuning; SmolVLM-Synthetic, the fine-tuned variant on synthetic data; and SmolVLM Instruct, the fine-tuned instruction variant, which can be used out of the box for interactive end-user applications. 
```

maybe, for better separation

File: smolvlm.md [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+
+In this blog post, we introduce SmolVLM, a new family of 2B small vision language models that can be used commercially and deployed to smaller local setups, with completely open training pipelines. 
+
+We release three models: SmolVLM-Base, which can be used for downstream fine-tuning, SmolVLM-Synthetic, the fine-tuned variant on synthetic data, and SmolVLM Instruct, the fine-tuned instruction variant, which can be used out of the box for interactive end-user applications. 
+
+This release comes with open-source models integrated into transformers, [a demo built on SmolVLM Instruct](https://huggingface.co/spaces/HuggingFaceTB/SmolVLM), and a supervised fine-tuning script. We have used the datasets previously used for Idefics3: [the Cauldron](https://huggingface.co/datasets/HuggingFaceM4/the_cauldron) and [Docmatix](https://huggingface.co/datasets/HuggingFaceM4/Docmatix), which are also fully open-source. [TODO: give links to all before release]
```
Comment (pcuenca) [inline smolvlm.md:27]: reminder: links and remove comment

File: smolvlm.md [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+
+# Table of Contents
```
Comment (pcuenca) [inline smolvlm.md:29]:
```suggestion
## Table of Contents
```

File: smolvlm.md [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+- Diagram understanding (AI2D)
+- General multimodal understanding (MMMU and MMStar). 
+
+To select the optimal checkpoint, we created a single metric by combining these benchmarks with different manually assigned weights to reflect their relative importance in assessing the model's capabilities. We used this single metric to select the best checkpoint. Generally, the models tended to do great on most benchmarks with more training, but their relative performance on DocVQA would decrease considerably. 
+
+## Fine-tuning
+
+You can fine-tune SmolVLM using transformers and apply alignment techniques using TRL 🚀
+
+We provide a notebook to fine-tune it on the VQAv2 dataset, optionally using  LoRA, QLoRA or full fine-tuning. In the notebook, you can find some tricks to save up even more memory and have a larger batch size to fit SmolVLM inside consumer GPUs, like L4, for training. With batch sizes of 4, 8-bit loading with QLoRA and gradient checkpointing we can fine-tune in L4, and it consumes around ~16 GBs of VRAM. This makes it possible to fine-tune your SmolVLM using Colab! You can play around with the parameters to get a nice point in training duration-memory trade-off.
```
Comment (merveenoyan) [inline smolvlm.md:237]:
```suggestion
We provide a [notebook](https://github.com/huggingface/smollm/tree/main/Smol_VLM_FT.ipynb) to fine-tune it on the VQAv2 dataset, optionally using  LoRA, QLoRA or full fine-tuning. In the notebook, you can find some tricks to save up even more memory and have a larger batch size to fit SmolVLM inside consumer GPUs, like L4, for training. With batch sizes of 4, 8-bit loading with QLoRA and gradient checkpointing we can fine-tune in L4, and it consumes around ~16 GBs of VRAM. This makes it possible to fine-tune your SmolVLM using Colab! You can play around with the parameters to get a nice point in training duration-memory trade-off. 
```

File: smolvlm.md [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+```
+A detailed tutorial on preference tuning vision-based LLM can be found here: [dpo_vlm](https://huggingface.co/blog/dpo_vlm).
+
+## Wrapping Up
+
+We introduced SmolVLM, a fully open, small, and mighty VLM for the community! We also provide tools for the community to use and customize it. We are looking forward to seeing what you will create with SmolVLM.
+
+Below are some resources if you would like to read more about all things related to SmolVLM.
+
+Start playing with SmolVLM using [this demo](https://huggingface.co/spaces/HuggingFaceTB/SmolVLM).
```
Comment (merveenoyan) [inline smolvlm.md:255]:
```suggestion
- Start playing with SmolVLM using [this demo](https://huggingface.co/spaces/HuggingFaceTB/SmolVLM).
```

File: smolvlm.md [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+Learn how to fine-tune SmolVLM on VQAv2 using [this notebook](https://github.com/merveenoyan/smol-vision/blob/main/Idefics_FT.ipynb )
```
Comment (merveenoyan) [inline smolvlm.md:256]:
```suggestion
- Learn how to fine-tune SmolVLM on VQAv2 using [this notebook](https://github.com/merveenoyan/smol-vision/blob/main/Idefics_FT.ipynb ) 
```

File: smolvlm.md [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+
+## Wrapping Up
+
+We introduced SmolVLM, a fully open, small, and mighty VLM for the community! We also provide tools for the community to use and customize it. We are looking forward to seeing what you will create with SmolVLM.
+
+Below are some resources if you would like to read more about all things related to SmolVLM.
+
+Start playing with SmolVLM using [this demo](https://huggingface.co/spaces/HuggingFaceTB/SmolVLM).
+Learn how to fine-tune SmolVLM on VQAv2 using [this notebook](https://github.com/merveenoyan/smol-vision/blob/main/Idefics_FT.ipynb ) 
+Learn more about [vision language models](https://huggingface.co/blog/vlms)
```
Comment (merveenoyan) [inline smolvlm.md:257]:
```suggestion
- Learn more about [vision language models](https://huggingface.co/blog/vlms)
```

File: smolvlm.md [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+
+### Travel Tips For Visiting The Grand Palace:
+* **Best Time To Visit**: The best time to visit would likely be early morning before sunrise when temperatures drop significantly making walking around easier.
+* **Transportation**: Public transportation options exist but they might not provide direct access to the Grand Palace itself; consider taking a taxi instead.
+* **Food Options**: There aren't many dining facilities available near the Grand Palace yet local street food vendors do offer delicious meals nearby.
+* **Clothing Guidelines**: While not mandatory, wearing traditional Thai clothing (sarongs) could add cultural immersion experience. However, shorts/trousers are allowed provided they don't go below knee level.
+* **Safety Measures**: Visitors must follow signs indicating restricted zones marked "No Photography" among others. Always carry identification documents just in case.
+
+By following these guidelines, your trip will undoubtedly enhance both your understanding of Thailand’s rich historical tapestry and personal enjoyment!
+    </td>
```
Comment (pcuenca) [inline smolvlm.md:80]: These might render weirdly because of the length (haven't checked)

**Review by merveenoyan — APPROVED** [review]
minor nits and added link to notebook, LGTM otherwise

File: smolvlm.md [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+  </tr>
+</table>
+
+## Architecture
+
+<img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/self_attention_architecture_smolvlm.png" width="800" height="auto" alt="Image description">
+
+For SmolVLM, we closely followed the architecture from Idefics3, to the point that we use the same implementation in transformers. There are, however a few key differences:
+ - We replaced Llama 3.1 8B with SmolLM2 1.7B as the language backbone.
+ - We more aggressively compress the patched information by reducing the information 9x using the pixel shuffle strategy, compared to 4x with idefics3.
```
Comment (pcuenca) [inline smolvlm.md:108]:
```suggestion
 - We more aggressively compress the patched visual information by reducing the information 9x using the pixel shuffle strategy, compared to 4x with idefics3.
```

(to separate from the first point)

File: smolvlm.md [hunk 15]
```diff
@@ -0,0 +1,257 @@
...
+ - We use patches of 384*384, instead of 364x364, because 384 is divisible by 3, which is necessary for our pixel shuffle strategy to work.
+ - For this, we change the vision backbone to use shape-optimized SigLIP with patches of 384x384 pixels and inner patches of 14x14.
+
+
+```
+All the models in this release are built on the shape-optimized SigLIP as image encoder and SmolLM2 for text decoder part, trained on datasets with Apache 2.0 license. We release the SmolVLM checkpoints under the Apache 2.0 license.
+```
```
Comment (pcuenca) [inline smolvlm.md:115]: I think this may be superfluous now, it's mentioned before. But ok to leave for stress if you prefer!

File: smolvlm.md [hunk 16]
```diff
@@ -0,0 +1,257 @@
...
+| InternVL2 2B      | 34.3       | 46.3                | 49.8         | 86.9          | 73.4          | 10.52                     |
+| PaliGemma 3B 448px| 34.9       | 28.7                | 48.3         | 32.2          | 56.0          | 6.72                      |
+| moondream2        | 32.4       | 24.3                | 40.3         | 70.5          | 65.2          | 3.87                      |
+| MiniCPM-V-2       | 38.2       | 39.8                | 39.1         | 71.9          | 74.1          | 7.88                      |
+| MM1.5 1B          | 35.8       | 37.2                | 0.0          | 81.0          | 72.5          | NaN                       |
+
+
+### Memory
+
+<img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/ram_smolvlm.png" width="900" height="auto" alt="Image description">
```
Comment (pcuenca) [inline smolvlm.md:135]:
```suggestion
<img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/ram_smolvlm.png" width="900" height="auto" alt="Inference GPU memory use for SmolVLM and other models">
```

File: smolvlm.md [hunk 17]
```diff
@@ -0,0 +1,257 @@
...
+
+SmolVLM provides the best memory usage among the existing suite of vision language models in transformers. This allows it to be efficiently run on-device, such as a laptop! Below, you can find the GPU memory usage in GBs for each model, running inference with one or two input images, and using the same images and text prompts in all tests. SmolVLM's efficiency in image encoding is built into the model. SmolVLM encodes each 384x384 image patch to 81 tokens. This results in SmolVLM encoding our test prompt and a single image in 1.2k tokens, whereas Qwen2-VL uses 16k tokens. This also explains why the memory consumption increases so much for 2 images with Qwen and InternVL. In contrast, the increase is much more moderate for SmolVLM and PaliGemma, which use a similar approach.
```
Comment (pcuenca) [inline smolvlm.md:137]:
```suggestion
SmolVLM provides the best memory usage among the existing suite of vision language models in transformers. This allows it to run efficiently on-device, such as a laptop! You can see above the GPU memory usage in GBs for each model, running inference with one or two input images, and using the same images and text prompts in all tests. SmolVLM's efficiency in image encoding is built into the model. SmolVLM encodes each 384x384 image patch to 81 tokens. This results in SmolVLM encoding our test prompt and a single image in 1.2k tokens, whereas Qwen2-VL uses 16k tokens. This also explains why the memory consumption increases so much for 2 images with Qwen and InternVL. In contrast, the increase is much more moderate for SmolVLM and PaliGemma, which use a similar approach.
```

File: smolvlm.md [hunk 18]
```diff
@@ -0,0 +1,257 @@
...
+
+SmolVLM also comes with TRL integration so you can apply Direct Preference Optimization (DPO) easily through the CLI. Get started by running `pip install trl accelerate peft` and then run the following command to fine-tune on [RLAIF-V] (https://huggingface.co/datasets/HuggingFaceH4/rlaif-v_formatted) dataset: 
+
+``` bash
+accelerate launch --config_file examples/accelerate_configs/multi_gpu.yaml examples/scripts/dpo_vlm.py  \\
+  --dataset_name HuggingFaceH4/rlaif-v_formatted --model_name_or_path HuggingFaceTB/SmolVLM-Instruct \\
+  --per_device_train_batch_size 8 --gradient_accumulation_steps 32 --dataset_num_proc 32 \\
+  --output_dir dpo_smolvlm_rlaif-v --bf16 --torch_dtype bfloat16 --use_peft --lora_target_modules=all-linear
```
Comment (pcuenca) [inline smolvlm.md:245]:
```suggestion
accelerate launch \
  --config_file examples/accelerate_configs/multi_gpu.yaml examples/scripts/dpo_vlm.py  \
  --dataset_name HuggingFaceH4/rlaif-v_formatted \
  --model_name_or_path HuggingFaceTB/SmolVLM-Instruct \
  --per_device_train_batch_size 8 \
  --gradient_accumulation_steps 32 \
  --dataset_num_proc 32 \
  --output_dir dpo_smolvlm_rlaif-v \
  --bf16 --torch_dtype bfloat16 \
  --use_peft --lora_target_modules=all-linear 
```

I don't think the `\\` would work, but I may be wrong.

**Review by pcuenca — APPROVED** [review]
🔥

File: smolvlm.md (hunk 13 above)
Comment (pcuenca) [reply smolvlm.md:80]: Images are too small :( 

Screenshot 2024-11-26 at 16 19 15

Comment (andimarafioti) [reply smolvlm.md:80]: yes, I noticed as well :/

File: smolvlm.md (hunk 4 above)
Comment (andimarafioti) [reply smolvlm.md:114]: looks good

**Pull request merged**